### PR TITLE
✨ (grapher) pull in data types from metadata

### PR DESCRIPTION
**Pull request by sophiamersmann**
> [!NOTE]  
> Wait for #3362 to be merged into master.

- Indicator metadata comes with types
- Up until now, Grapher has been ignoring type information and marked every column as mixed type
- This PR pulls in the metadata type information into Grapher by setting the appropriate column type

---

- ~Open question: Should we split `Numeric` into `Integer` and `Float`?~ Using `Integer` for `int` and `Numeric` for `float`

Closes https://github.com/owid/owid-grapher/issues/1692

### CODE REVIEW

File: packages/@ourworldindata/grapher/src/core/LegacyToOwidTable.ts [hunk 1]
```diff
@@ -551,13 +551,12 @@ const variableTypeToColumnType = (type: OwidVariableType): ColumnTypeNames => {
...
-        // case "string":
-        //     return ColumnTypeNames.String
-        // case "float":
-        // case "int":
-        //     return ColumnTypeNames.Numeric
-        // case "mixed":
+        case "string":
+            return ColumnTypeNames.String
+        case "float":
+        case "int":
```
Comment (marcelgerber) [inline packages/@ourworldindata/grapher/src/core/LegacyToOwidTable.ts:557]: We already have `ColumnTypeNames.Integer`, which I think will be a good fit for `"int"`.

The main difference to `Numeric` is that it formats numbers with 0 decimal places by default.

**Review by marcelgerber — APPROVED** [review]

File: packages/@ourworldindata/grapher/src/core/LegacyToOwidTable.ts (hunk 1 above)
Comment (sophiamersmann) [reply packages/@ourworldindata/grapher/src/core/LegacyToOwidTable.ts:557]: Sounds good! Do you think we should use `NumericColumn` for float types directly, or should we add another `FloatColumn` that inherits from `NumericColumn`?

Comment (marcelgerber) [reply packages/@ourworldindata/grapher/src/core/LegacyToOwidTable.ts:557]: I think it's fine to just use `NumericColumn`.

It's not like a `float` type is _only_ floats - there'll be integer values in these columns, too.

Comment (sophiamersmann) [reply packages/@ourworldindata/grapher/src/core/LegacyToOwidTable.ts:557]: Makes sense!

**Comment by sophiamersmann** [comment]
After marking int columns as `Integer` columns in Grapher, the SVG reported a few charts with differences like this:

Screenshot 2024-03-28 at 12 23 42

The incorrect rounding stems from the fact that the data column is marked as integer (which is technically correct), but then a conversion factor is applied to it, which turns the Integer column into a Non-integer column.

This is fixed by updating the column definition when a conversion factor is applied.

---
For the remaining SVG differences, integer rounding is an improvement
- http://staging-site-data-types/grapher/bcg-immunization-coverage-for-tb-among-1-year-olds
- http://staging-site-data-types/grapher/endemic-bird-species-by-country
- http://staging-site-data-types/grapher/endemic-reef-forming-coral-species
- http://staging-site-data-types/grapher/threatened-endemic-coral

**Comment by marcelgerber** [comment]
In theory, we could check if the `conversionFactor` is itself integer (for example `1000`), and in that case keep the column type as `Integer`.
Not necessary, though, the difference is not very stark.

File: packages/@ourworldindata/grapher/src/core/LegacyToOwidTable.ts [hunk 2]
```diff
@@ -145,6 +146,14 @@ export const legacyToOwidTableAndDimensions = (
             values = values.map((value) =>
                 isNumber(value) ? value * conversionFactor : value
             )
+
+            // If a non-int conversion factor is applied to an integer column,
+            // we end up with a numeric column.
+            if (
+                valueColumnDef.type === ColumnTypeNames.Integer &&
+                isInteger(conversionFactor)
```
Comment (marcelgerber) [inline packages/@ourworldindata/grapher/src/core/LegacyToOwidTable.ts:154]:
```suggestion
                !isInteger(conversionFactor)
```

**Review by marcelgerber — CHANGES_REQUESTED** [review]

File: packages/@ourworldindata/grapher/src/core/LegacyToOwidTable.ts (hunk 2 above)
Comment (sophiamersmann) [reply packages/@ourworldindata/grapher/src/core/LegacyToOwidTable.ts:154]: opps

**Pull request merged**